core/server/src/main/java/alluxio/master/file/meta/Inode.java

```diff
@@ -277,24 +292,93 @@ public T setOwner(String owner) {
   protected abstract T getThis();
 
   /**
-   * Acquires the read lock for this inode.
+   * Obtains a read lock on the inode. This should only be used when the inode is accessed by ID
+   * and not by path or parent.
    */
   public void lockRead() {
     mLock.readLock().lock();
   }
 
   /**
-   * Releases the read lock for this inode.
+   * Obtains a read lock on the inode. Afterward, checks the inode state to ensure the full inode
+   * path is consistent with what the caller is expecting. If the state is inconsistent, an
+   * exception will be thrown and the lock will be released.
+   *
+   * @param parent the expected parent inode
+   * @param name the expected name of the inode to be locked
+   * @throws InvalidPathException if the parent and/or name is not as expected
+   */
+  public void lockReadAndCheckFullPath(Inode parent, String name) throws InvalidPathException {
```

Comment:
does this really check the full path? what if the parent of the parent is renamed? the full path has changed but this method will not throw an exception